test_runner/regress/test_tenant_detach.py

```diff
@@ -108,6 +108,59 @@ def test_tenant_reattach(
         assert env.pageserver.log_contains(".*download.*failed, will retry.*")
 
 
+# Detach and re-attach a tenant with a branch
+#
+# This exercises a race condition after tenant attach, where the
+# branch point on the ancestor timeline is beyond the ancestor's
+# last-record LSN. GetPage incorrectly followed the timeline to the
+# ancestor without waiting for the missing WAL to arrive.
+#
+# That could lead to permanent damage in the child timeline, because
+# ingesting the WAL on the child timeline depended on incorrect view
+# of the parent. This test reproduced one such case; the symptom was a
+# 'FATAL: database "neondb" does not exist' error on the child, when
+# trying to connect after re-attaching the tenant. The WAL record of
+# CREATE DATABASE was missed when searching the DbDirectory key in the
+# ancestor.
+@pytest.mark.parametrize("remote_storage_kind", available_remote_storages())
+def test_tenant_with_branches_reattach(
+    neon_env_builder: NeonEnvBuilder,
+    remote_storage_kind: RemoteStorageKind,
+):
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_tenant_reattach",
+    )
+
+    env = neon_env_builder.init_start()
+    pageserver_http = env.pageserver.http_client()
+
+    # create new nenant
+    tenant_id, timeline_id = env.neon_cli.create_tenant()
+
+    main_endpoint = env.endpoints.create_start("main", tenant_id=tenant_id)
+    with main_endpoint.connect().cursor() as cur:
+        cur.execute("CREATE DATABASE neondb")
+        current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
+    env.neon_cli.create_branch("child", tenant_id=tenant_id, ancestor_start_lsn=current_lsn)
+
+    with main_endpoint.connect(dbname="neondb").cursor() as cur:
```

Comment:
Python technical question: does this mean both of the `db connection` (from `connect(dbname=...)` and `cursor` are context objects which will be cleared on exit from `with` or will this leak the connections?

Reply:
I'm not sure, but I realized that the `cursor` function can also take connection parameters as arguments. So I removed the connect() calls, making the question moot.

(In general I'm not to worried about leaking things during tests. They get cleaned up at end of test anyway)

Reply:
Yeah for connections it is not important, but if you had chained the `endpoint.start()` (or something which provides that) and another method on that, it'd be interesting to know if this should be an antipattern none of our tools alert on or not. Leaking a compute would result in unwanted behaviour most likely.